modules/3d/src/octree.cpp

```diff
@@ -0,0 +1,507 @@
+        return deletePointRecurse(node);
+    }
+    else
+    {
+        return false;
+    }
+}
+
+bool deletePointRecurse(Ptr<OctreeNode>& node)
+{
+    if(node == nullptr)
+        return false;
+    if(node->isLeaf)
+    {
+        if( !node->pointList.empty())
+        {
+            Ptr<OctreeNode> parent = node->parent;
+            parent->children[node->parentIndex] = nullptr;
+            node.release();
+
+            return deletePointRecurse(parent);
+        }
+        else
+        {
+            return true;
+        }
+    }
+    else
+    {
+        bool deleteFlag = true;
+
+        // Only all children was deleted, can we delete the tree node.
+        for(size_t i = 0; i< octreeChildNum; i++)
+        {
+            if(node->children[i] != nullptr)
+            {
+                deleteFlag = false;
+                break;
+            }
+        }
+
+        if(deleteFlag)
+        {
+            Ptr<OctreeNode> parent = node->parent;
+            node = nullptr;
+            return deletePointRecurse(parent);
+        }
+        else
+        {
+            return true;
+        }
+    }
+}
+
+void insertPointRecurse( Ptr<OctreeNode>& node,  const Point3f& point, int maxDepth)
+{
+    if(!node->isPointInBound(point, node->origin, node->size))
+    {
+        CV_Error(Error::StsBadArg, "The point is out of boundary!");
+    }
+
+    if(node->depth == maxDepth)
+    {
+        node->isLeaf = true;
+        node->pointList.push_back(point);
+        return;
+    }
+
+    double childSize = node->size / 2.0;
+    size_t xIndex = point.x < ( node->origin.x + float(childSize) + epsilon ) ? 0 : 1;
+    size_t yIndex = point.y < ( node->origin.y + float(childSize) + epsilon ) ? 0 : 1;
+    size_t zIndex = point.z < ( node->origin.z + float(childSize) + epsilon ) ? 0 : 1;
+    size_t childIndex = xIndex + yIndex * 2 + zIndex * 4;
+
+    if(node->children[childIndex] == nullptr)
+    {
+        Point3f childOrigin = node->origin + Point3f(xIndex * float(childSize), yIndex * float(childSize), zIndex * float(childSize));
+        node->children[childIndex] = new OctreeNode(node->depth + 1, childSize, childOrigin, int(childIndex));
+        node->children[childIndex]->parent = node;
+    }
+    insertPointRecurse(node->children[childIndex], point, maxDepth);
+}
+
+// For Nearest neighbor search.
+template<typename T>
+struct PQueueElem
+{
+    PQueueElem() : dist(0), t(0) {}
+    PQueueElem(float _dist, T _t) : dist(_dist), t(_t) {}
+    float dist;
+    T t;
+
+    bool
+    operator<(const PQueueElem<T> p1) const
+    {
+        return (this->dist < p1.dist);
+    }
+};
+
+float SquaredDistance(const Point3f& query, const Point3f& origin)
+{
+    float x = query.x - origin.x;
+    float y = query.y - origin.y;
+    float z = query.z - origin.z;
+    return x * x + y * y + z * z;
+}
+
+bool overlap(const Ptr<OctreeNode>& node, const Point3f& query, float squareRadius)
+{
```

Comment:
> `const Ptr<OctreeNode>& node`

We don't need smart pointer here (with many dereference operations).
Pass `const OctreeNode& node` instead.

The similar note is for other applicable functions (not all).

Reply:
Thanks, fixed.